micropip/transaction.py

```diff
@@ -36,14 +36,31 @@ class Transaction:
     failed: list[Requirement] = field(default_factory=list)
 
     verbose: bool | int | None = None
+    constraints: list[str] | None = None
 
-    def __post_init__(self):
+    def __post_init__(self) -> None:
         # If index_urls is None, pyodide-lock.json have to be searched first.
         # TODO: when PyPI starts to support hosting WASM wheels, this might be deleted.
         self.search_pyodide_lock_first = (
             self.index_urls == package_index.DEFAULT_INDEX_URLS
         )
 
+        self.constrained_reqs: dict[str, Requirement] = {}
+
+        for constraint in self.constraints or []:
+            con = Requirement(constraint)
+            if not con.name:
+                logger.debug("Transaction: discarding nameless constraint: %s", con)
+                continue
+            if con.extras:
+                logger.debug("Transaction: discarding [extras] constraint: %s", con)
+                continue
+            if not (con.url or len(con.specifier)):
+                logger.debug("Transaction: discarding versionless constraint: %s", con)
+                continue
+            con.name = canonicalize_name(con.name)
+            self.constrained_reqs[con.name] = con
```

Comment:
Could you split this into a separate function and add some test?

Reply:
moved to `_utils`